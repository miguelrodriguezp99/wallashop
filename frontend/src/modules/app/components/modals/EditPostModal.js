import {
	Button,
	Checkbox,
	Input,
	Modal,
	ModalBody,
	ModalContent,
	ModalFooter,
	ModalHeader,
	useDisclosure,
} from "@nextui-org/react";
import React, { useState } from "react";
import { RiEditBoxLine } from "react-icons/ri";
import { toast } from "sonner";
import { editPost } from "../../../../backend/postService.js";
import Selector from "../selectors/Selector.js";
import "./../../styles/navbar.css";

export default function EditPostModal({ post }) {
	// ------ Context -----------------------------------------------
	const { isOpen, onOpenChange, onOpen } = useDisclosure();

	const [postTitle, setPostTitle] = useState(post?.title);
	const [postDescription, setPostDescription] = useState(post?.description);
	const [postUrl, setPostUrl] = useState(post?.url);
	const [postPrice, setPostPrice] = useState(post?.price);
	const [postCategory, setPostCategory] = useState(post?.category?.name);
	const [isActive, setIsActive] = useState(post?.active);

	const onSuccess = () => {
		toast.success("Post editado.");
		window.location.reload(true);
	};

	const onErrors = () => {
		toast.error("Error al editar el post.");
	};

	const handleSubmit = () => {
		const postDto = {
			title: postTitle,
			description: postDescription,
			url: postUrl,
			price: postPrice,
			images: post.images,
			active: isActive,
		};

		console.log(postDto);

		editPost(post.id, postDto, onSuccess, onErrors);
	};

	return (
		<>
			<Button
				onClick={onOpen}
				className="mt-6 border border-green-600 rounded text-white bg-[#18181B] hover:bg-green-500"
			>
				<RiEditBoxLine size={20} />
			</Button>

			<Modal
				isOpen={isOpen}
				onOpenChange={onOpenChange}
				className="w-full"
				scrollBehavior="inside"
				backdrop="blur"
			>
				<ModalContent className="bg-[#18181B] w-full mt-10">
					{(onClose) => (
						<>
							<ModalHeader className="flex flex-col gap-1 text-white">
								{" "}
								Editar el post{" "}
							</ModalHeader>
							<ModalBody className="w-full" >
								<Input 
									size="sm"
									type="text"
									label="Title"
									value={postTitle}
									onChange={(e) => setPostTitle(e.target.value)}
									isRequired
								/>

								<Input
									size="sm"
									type="text"
									label="Description"
									value={postDescription}
									onChange={(e) => setPostDescription(e.target.value)}
									isRequired
								/>
								<Input
									size="sm"
									type="text"
									label="URL"
									value={postUrl}
									onChange={(e) => setPostUrl(e.target.value)}
									isRequired
								/>

								<Input
									type="number"
									value={postPrice}
									onChange={(e) => setPostPrice(parseFloat(e.target.value))}
									placeholder={postPrice}
									labelPlacement="outside"
									isRequired
									startContent={
										<div className="pointer-events-none flex items-center">
											<span className="text-default-400 text-small">â‚¬</span>
										</div>
									}
								/>
								<Selector
									value={postCategory}
									fn={setPostCategory}
									placeholder={post.category.name}
									isRequired
								/>
								<div className="flex flex-col gap-2">
									<Checkbox isSelected={isActive} onValueChange={setIsActive} />
									<p className="text-default-500">
										Post Status: {isActive ? "Active" : "Inactive"}
									</p>
								</div>
							</ModalBody>

							<ModalFooter>
								<Button color="danger" variant="light" onPress={onClose}>
									Close
								</Button>
								<Button color="primary" onPress={handleSubmit}>
									Submit
								</Button>
							</ModalFooter>
						</>
					)}
				</ModalContent>
			</Modal>
		</>
	);
}
