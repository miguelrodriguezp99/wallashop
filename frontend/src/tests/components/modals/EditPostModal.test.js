import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import EditPostModal from './../../../modules/app/components/modals/EditPostModal';

describe('EditPostModal component', () => {
  const mockPost = {
    id: 1,
    title: 'Sample Title',
    description: 'Sample Description',
    url: 'https://example.com',
    price: 20.99,
    category: {
      id: 1,
      name: 'Sample Category',
    },
    active: true,
  };

  it('renders correctly', () => {
    const tree = renderer
      .create(
        <MemoryRouter>
          <EditPostModal post={mockPost} />
        </MemoryRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});