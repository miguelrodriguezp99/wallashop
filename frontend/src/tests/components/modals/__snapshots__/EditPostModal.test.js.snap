// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditPostModal component renders correctly 1`] = `
<button
  className="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-unit-4 min-w-unit-20 h-unit-10 text-small gap-unit-2 [&>svg]:max-w-[theme(spacing.unit-8)] data-[pressed=true]:scale-[0.97] transition-transform-colors-opacity motion-reduce:transition-none data-[hover=true]:opacity-hover mt-6 border border-green-600 rounded text-white bg-[#18181B] hover:bg-green-500"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onDragStart={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchCancel={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  type="button"
>
  <svg
    fill="currentColor"
    height={20}
    stroke="currentColor"
    strokeWidth="0"
    style={
      Object {
        "color": undefined,
      }
    }
    viewBox="0 0 24 24"
    width={20}
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M16.7574 2.99666L14.7574 4.99666H5V18.9967H19V9.2393L21 7.2393V19.9967C21 20.5489 20.5523 20.9967 20 20.9967H4C3.44772 20.9967 3 20.5489 3 19.9967V3.99666C3 3.44438 3.44772 2.99666 4 2.99666H16.7574ZM20.4853 2.09717L21.8995 3.51138L12.7071 12.7038L11.2954 12.7062L11.2929 11.2896L20.4853 2.09717Z"
    />
  </svg>
</button>
`;
